### PR TITLE
Parallel shader cache loading

### PR DESCRIPTION
**Pull request by Dravonic**
Use all available threads during the initial shader compilation process in hopes of speeding it up.

The code is mostly unchanged and has multi-threading essentially hammered into it. The result isn't exactly pretty but hey, it works.

Multi-threaded path in Vulkan only, at least for now.

### CODE REVIEW

**Comment by Zangetsu38** [comment]
i hope dev can set same for loading spu cache lol is very slow for now

**Comment by theoldsport** [comment]
Wew that was FAST

**Comment by raven02** [comment]
It is quick however i got two bar like this.

untitled

**Comment by theoldsport** [comment]
My bad, I wasn't paying attention,

_Loading pipeline objects_
**Master** - 8 seconds
**PR** - 2 seconds 

_Compiling pipeline objects_
**Master** - 33 seconds?
**PR** - instantly

Hard to test, because I know when I update builds, on first run, loading precompiled shaders takes longer

**Comment by Illynir** [comment]
On Nvidia it's the GLCache in the driver. The GLCache becomes invalid if one changes exe or after a reboot sometimes. If it is longer first it is simply that the driver recreates the GlCache. I don't think this PR will change anything because it's managed by the driver.

**Comment by Zangetsu38** [comment]
Yes but this PR very boost speed load shader for reboot and re lunch so

File: rpcs3/Emu/RSX/Common/ProgramStateCache.h [hunk 1]
```diff
@@ -149,7 +151,8 @@ class program_state_cache
 			return std::forward_as_tuple(I->second, true);
 		}
 		LOG_NOTICE(RSX, "VP not found in buffer!");
-		vertex_program_type& new_shader = m_vertex_shader_cache[rsx_vp];
+
```
Comment (kd-11) [inline rpcs3/Emu/RSX/Common/ProgramStateCache.h]: Don't add unnecessary formatting. If it doesn't change functionality, leave it alone

File: rpcs3/Emu/RSX/Common/ProgramStateCache.h [hunk 2]
```diff
@@ -169,6 +172,7 @@ class program_state_cache
 		std::memcpy(fragment_program_ucode_copy, rsx_fp.addr, fragment_program_size);
 		RSXFragmentProgram new_fp_key = rsx_fp;
 		new_fp_key.addr = fragment_program_ucode_copy;
+
```
Comment (kd-11) [inline rpcs3/Emu/RSX/Common/ProgramStateCache.h]: Same here

File: rpcs3/Emu/RSX/Common/ProgramStateCache.h [hunk 3]
```diff
@@ -305,11 +309,14 @@ class program_state_cache
 		LOG_NOTICE(RSX, "*** vp id = %d", vertex_program.id);
 		LOG_NOTICE(RSX, "*** fp id = %d", fragment_program.id);
 
-		m_storage[key] = backend_traits::build_pipeline(vertex_program, fragment_program, pipelineProperties, std::forward<Args>(args)...);
+		pipeline_storage_type pipeline = backend_traits::build_pipeline(vertex_program, fragment_program, pipelineProperties, std::forward<Args>(args)...);
+		s_mtx.lock();
```
Comment (kd-11) [inline rpcs3/Emu/RSX/Common/ProgramStateCache.h]: Just use RAII. Will also serialize the logger broadcast below

File: rpcs3/Emu/RSX/Overlays/overlays.h [hunk 4]
```diff
@@ -937,6 +937,18 @@ namespace rsx
 
 				return CELL_OK;
 			}
+
+			s32 progress_bar_set_limit(u32 index, u32 limit) {
```
Comment (kd-11) [inline rpcs3/Emu/RSX/Overlays/overlays.h]: brace on newline

File: rpcs3/Emu/RSX/rsx_cache.h [hunk 5]
```diff
@@ -432,36 +433,39 @@ namespace rsx
...
 
 			virtual void close()
 			{}
 		};
 
 		shaders_cache(backend_storage& storage, std::string pipeline_class, std::string version_prefix_str = "v1")
-			: version_prefix(version_prefix_str)
-			, pipeline_class_name(pipeline_class)
-			, m_storage(storage)
+		    : version_prefix(version_prefix_str)
```
Comment (kd-11) [inline rpcs3/Emu/RSX/rsx_cache.h]: unnecessary reformating (clang-format issues?)

File: rpcs3/Emu/RSX/rsx_cache.h [hunk 6]
```diff
@@ -432,36 +433,39 @@ namespace rsx
...
+		    , pipeline_class_name(pipeline_class)
+		    , m_storage(storage)
 		{
 			root_path = Emu.GetCachePath() + "/shaders_cache";
 		}
 
 		template <typename... Args>
-		void load(progress_dialog_helper* dlg, Args&& ...args)
+		void load(progress_dialog_helper* dlg, Args&&... args)
```
Comment (kd-11) [inline rpcs3/Emu/RSX/rsx_cache.h]: same here

File: rpcs3/Emu/RSX/rsx_cache.h [hunk 7]
```diff
@@ -505,39 +513,97 @@ namespace rsx
...
 
-			const auto prefix_length = version_prefix.length();
-			u32 processed = 0;
-			while (root.read(tmp) && !Emu.IsStopped())
+			// Preload everything needed to compile the shaders
+			// Can probably be parallelized too, but since it's mostly reading files it's probably not worth it
+			std::vector<std::tuple<pipeline_storage_type, RSXVertexProgram, RSXFragmentProgram>> unpackeds;
+			std::chrono::milliseconds last_update;
+			u32 processed_since_last_update = 0;
+			for (u32 i = 0; (i < entry_count) && !Emu.IsStopped(); i++)
```
Comment (kd-11) [inline rpcs3/Emu/RSX/rsx_cache.h]: i'd prefer to add blank line before this block. The declaration block above is large enough and has comments anyway

File: rpcs3/Emu/RSX/rsx_cache.h [hunk 8]
```diff
@@ -505,39 +513,97 @@ namespace rsx
...
+				{
+					dlg->update_msg(0, i + 1, entry_count);
+					dlg->inc_value(0, processed_since_last_update);
+					last_update = now;
+					processed_since_last_update = 0;
+				}
+			}
+
+			atomic_t<u32> processed(0);
+			std::function<void(u32)> shader_comp_worker = [&](u32 index) {
```
Comment (kd-11) [inline rpcs3/Emu/RSX/rsx_cache.h]: brace on newline since companion brace is also on newline

File: rpcs3/Emu/RSX/rsx_cache.h [hunk 9]
```diff
@@ -505,39 +513,97 @@ namespace rsx
...
+
 				auto unpacked = unpack(*(pipeline_data*)bytes.data());
-				m_storage.add_pipeline_entry(std::get<1>(unpacked), std::get<2>(unpacked), std::get<0>(unpacked), std::forward<Args>(args)...);
+				m_storage.preload_programs(std::get<1>(unpacked), std::get<2>(unpacked));
+				unpackeds.push_back(unpacked);
 
-				tally += delta;
-				if (tally > 1.f)
+				// Only update the screen at about 10fps since updating it everytime slows down the process
+				std::chrono::milliseconds now = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
```
Comment (kd-11) [inline rpcs3/Emu/RSX/rsx_cache.h]: Pretty sure you dont need time since epoch, just a forward incrementing counter. Consider just using steady_clock::now()

File: rpcs3/Emu/RSX/rsx_cache.h [hunk 10]
```diff
@@ -505,39 +513,97 @@ namespace rsx
...
+					dlg->inc_value(0, processed_since_last_update);
+					last_update = now;
+					processed_since_last_update = 0;
+				}
+			}
+
+			atomic_t<u32> processed(0);
+			std::function<void(u32)> shader_comp_worker = [&](u32 index) {
+				u32 pos;
+				while (((pos = ++processed - 1) < entry_count) && !Emu.IsStopped())
```
Comment (kd-11) [inline rpcs3/Emu/RSX/rsx_cache.h]: Why not `pos = processed++`? atomic_t has a post incrementer op.

**Comment by theoldsport** [comment]
Updated my test, looks about right now 👍

File: rpcs3/Emu/RSX/rsx_cache.h [hunk 11]
```diff
@@ -505,39 +513,97 @@ namespace rsx
...
 				{
-					u32 value = (u32)tally;
-					dlg->inc_value(value);
+					auto unpacked = unpackeds[pos];
+					m_storage.add_pipeline_entry(std::get<1>(unpacked), std::get<2>(unpacked), std::get<0>(unpacked), std::forward<Args>(args)...);
 
-					tally -= (f32)value;
+					dlg->update_msg(1, processed.load(), entry_count);
```
Comment (kd-11) [inline rpcs3/Emu/RSX/rsx_cache.h]: You should also do the timed updates for OGL. I was going to add this at some point but then other stuff got in the way. A huge chunk of the performance bump comes from the timed update that does not invoke the vblank wait in the driver especially on nvidia where frametimes are not uncapped by default.

**Comment by Megolas** [comment]
If you want, you can batch the pipeline loads, too, and massively increase loading speed, tested it with shit code a while ago

**Comment by kd-11** [comment]
Does this work with qt dialog? It seems the normal qt dialog is not updated for this. Its interchangeable with the native interface in terms of methods.

**Comment by Dravonic** [comment]
Didn't even consider the normal dialogs. Turns out they aren't completely broken but the progress bar doesn't work properly. I'll try to fix it up.
